Add BigQuery Spring Integration File Message Handler

dzou opened the pull request:
This adds a Spring Integration `MessageHandler` implementation for data files for BigQuery load operations. This is the outbound channel adapter for loading data files into big query.

Contributes to #1764.

meltsufin commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,135 @@
+import org.springframework.util.StreamUtils;
+
+/**
+ * Outbound channel adapter which handles sending and loading files to a BigQuery table.
```
Please include author/version.

artembilan replied:
It's not a *channel adapter*. It is really a `MessageHandler`. Please, be more specific in JavaDocs.

meltsufin commented on spring-cloud-gcp-bigquery/pom.xml:
```diff
@@ -0,0 +1,39 @@
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>spring-cloud-gcp-bigquery</artifactId>
```
Can you include artifact name and description tags?

meltsufin commented on spring-cloud-gcp-bigquery/pom.xml:
```diff
@@ -0,0 +1,39 @@
+    <groupId>org.springframework.cloud</groupId>
+    <version>1.2.0.BUILD-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
```
We usually put this line before the `<parent>` tag.

meltsufin commented on spring-cloud-gcp-bigquery/pom.xml:
```diff
@@ -0,0 +1,39 @@
+  ~ you may not use this file except in compliance with the License.
+  ~ You may obtain a copy of the License at
+  ~
+  ~      https://www.apache.org/licenses/LICENSE-2.0
```
License comment not required for `pom.xml`.

meltsufin commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,135 @@
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.bigquery.integration;
```
Maybe change to `...integration.outbound` for consistency with Pub/Sub module?

meltsufin commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,135 @@
+	@Override
+	protected Object handleRequestMessage(Message<?> message) {
+		try {
+			String datasetName = message.getHeaders().get(BigQuerySpringMessageHeaders.DATASET_NAME, String.class);
```
Interesting choice to use the headers for this rather than properties of the actual channel adapter.
Maybe we can use both, where headers take precedence if set.
I'm curious @artembilan's opinion on this.

artembilan replied:
See `PubSubMessageHandler` for its `topicExpression` as a sample how to be.

It is indeed must not be a strict requirement for just header population: we can obtain required info from any other places. That's where a SpEL with its evaluation context comes to the help. It helps us to avoid header-enricher upfront.

dzou replied:
Done. I've changed it so the user can set a default for the message handler options; however the message can provide headers which will override the default settings.

meltsufin commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,135 @@
+				sink.close();
+			}
+
+			return writer.getJob();
```
Should we block waiting for the job to complete?
Is this technically a "reply"?
@artembilan

artembilan replied:
Why do we worry about this `Job` as a reply?

dzou replied:
Seems related to this: https://github.com/spring-cloud/spring-cloud-gcp/pull/1822#discussion_r314447943

will have conversation there.

meltsufin commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQuerySpringMessageHeaders.java:
```diff
@@ -0,0 +1,38 @@
+
+/**
+ * Spring Integration {@link org.springframework.messaging.Message} headers used with
+ * Spring Cloud GCP BigQuery integration.
```
author / version

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,135 @@
+	private WriteDisposition writeDisposition = WriteDisposition.WRITE_APPEND;
+
+	public BigQueryFileMessageHandler(BigQuery bigQuery) {
+		this.bigQuery = bigQuery;
```
`Assert.notNull()`?

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,135 @@
+			return writer.getJob();
+		}
+		catch (FileNotFoundException e) {
+			throw new MessageHandlingException(message, "Failed to find file to write to BigQuery.", e);
```
Please, don't use `MessageHandlingException` if you don't include `this` into an exception message: https://github.com/spring-projects/spring-integration/issues/2748

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,135 @@
+	 * BigQuery tables.
+	 * @param writeDisposition whether data should be appended or truncated to the BigQuery table
+	 */
+	public void setWriteDisposition(WriteDisposition writeDisposition) {
```
Setters (and getters for them) must go just after constructor(s), not before `proptected` methods with logic.

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQuerySpringMessageHeaders.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public final class BigQuerySpringMessageHeaders {
+
+	private static final String PREFIX = "gcp_bigquery_";
```
What is wrong to have this as `public` as well?
Imagine I want to have a logic to filter or map all the headers which match a `BigQuerySpringMessageHeaders.PREFIX + '*'` pattern...

artembilan commented on spring-cloud-gcp-bigquery/src/test/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryFileMessageHandlerIntegrationTests.java:
```diff
@@ -0,0 +1,94 @@
+				new MessageHeaders(messageHeaders));
+
+		Job job = (Job) this.messageHandler.handleRequestMessage(message);
+		job.waitFor();
```
I think we need to have some logic in the `BigQueryFileMessageHandler` to handle this `Job` internally. Something like `sync` and `sendTimeout` options and some async handling otherwise.
See https://github.com/spring-projects/spring-integration-kafka/blob/master/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java for more info.

dzou replied:
OK. I have a question though -

This `Job` object is something returned by the Google Client libs.

If you want to wait for the job to complete **sync**, you call:
```
job.waitFor()

// Then you can start using the object
job.getResults().getRowsUpdated()
```

If you want to do this **async**, you have to do some kind of polling like this:
```
while (!JobStatus.State.DONE.equals(job.getStatus().getState())) {
  Thread.sleep(1000L);
  job = job.reload();
}

// Then you can start using the object
job.getResults().getRowsUpdated()
```

What is the recommended practice in Spring Integration for abstracting polling logic from the user?

artembilan replied:
Well, it still sounds like a background task and a `ListenableFuture` as a return from this `handleRequestMessage`.
A `SettableListenableFuture` should be used internally for populating the value.

dzou replied:
Gotcha sounds good. In that case, would I also need to add a thread pool responsible for polling these jobs? Or would the user provide this?

artembilan replied:
We can have a default one, but still allow end-user to inject a managed one. We don't have such a sample though to share because we have always relied on the async execution in the target client, but essentially, it doesn't matter for end-user who is responsible for task execution.

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -36,30 +36,97 @@
+			FormatOptions formatOptions =
+					message.getHeaders().containsKey(BigQuerySpringMessageHeaders.FORMAT_OPTIONS)
+							? message.getHeaders().get(
+									BigQuerySpringMessageHeaders.FORMAT_OPTIONS, FormatOptions.class)
```
Is really `FormatOptions` a thing to be carried in headers?
Is it really so message-specific?
Is it JSON-, Serialization-compatible?

dzou replied:
Is really FormatOptions a thing to be carried in headers?
- Hmm, I am considering the use-case where somebody might have a couple different file formats they wish to load to the database; imagine having both CSV and JSON. In that case, I thought it would be reasonable to let the user specify the file format in the message header.

Is it really so message-specific?
- Based on the above reasoning if correct then it would be message specific.

Is it JSON-, Serialization-compatible?
- It [implements Serializable](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/FormatOptions.java), however unsure about the JSON thing.

artembilan replied:
Well, I think that existing ctor should make it compatible with JSON (de)serialization even if it is a package protected.
So, this is fine with me so far.

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQuerySpringMessageHeaders.java:
```diff
@@ -14,15 +14,19 @@
 public final class BigQuerySpringMessageHeaders {
 
-	private static final String PREFIX = "gcp_bigquery_";
+	/** BigQuery Spring Cloud GCP message header prefix. */
```
That's not a proper JavaDoc format.
Please, make it really multi-lined

dzou commented:
> And general question: what is wrong with SpEL expressions suggestion?

Nothing wrong with it - however I thought it would just add some unneeded complexity as I imagined the dataset name and table names would just be static strings; these are the datasets/tables that the user loads their data into. But if the static assumption is incorrect then I will add SpEL.

artembilan commented:
> But if the static assumption is incorrect then I will add SpEL.

It's definitely not correct.
See `MongoDbStoringMessageHandler` and its `collectionNameExpression`. Or `RedisStoreWritingMessageHandler` and its `keyExpression`.
BTW, default might be really expressed as values from headers, meanwhile user-defined static values also can be expressed as expressions: `ValueExpression` or `LiteralExpression`.

However this job can be done in the separate PR. What you have so far is OK to go.

dzou commented:
> > But if the static assumption is incorrect then I will add SpEL.
> 
> It's definitely not correct.
> See `MongoDbStoringMessageHandler` and its `collectionNameExpression`. Or `RedisStoreWritingMessageHandler` and its `keyExpression`.
> BTW, default might be really expressed as values from headers, meanwhile user-defined static values also can be expressed as expressions: `ValueExpression` or `LiteralExpression`.
> 
> However this job can be done in the separate PR. What you have so far is OK to go.

In the latest commit I added SpEL expressions for the table name and dataset name. Also added support for the `sync` option; the handler now has polling and will return listenableFuture on completion of the load job.

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,279 @@
+				message.getHeaders().containsKey(BigQuerySpringMessageHeaders.FORMAT_OPTIONS)
+						? message.getHeaders().get(
+								BigQuerySpringMessageHeaders.FORMAT_OPTIONS, FormatOptions.class)
+						: this.formatOptions;
```
Why this one doesn't go to the expression?

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,279 @@
+								BigQuerySpringMessageHeaders.FORMAT_OPTIONS, FormatOptions.class)
+						: this.formatOptions;
+
+		TableId tableId = TableId.of(datasetName, tableName);
```
I think we need to have an `Assert.notNull()` for these options.

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,279 @@
+							new RuntimeException(job.getStatus().getError().getMessage()));
+				}
+				else {
+					result.set(job);
```
What is the business purpose of this `Job`? Why do we really need it in comparison of sent file?

dzou replied:
So Job is useful because the user can get statistics and metadata from the load operation out of it. For example # of successfully inserted rows, # of unsuccessfully loaded rows, time taken, etc.

https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/bigquery/Job.html

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,279 @@
+	}
+
+	@Override
+	protected ListenableFuture handleRequestMessage(Message<?> message) {
```
You should consider do not return `ListenableFuture` in the `sync` mode.

So, just `return result.get(this.timeout.getSeconds(), TimeUnit.SECONDS);` should be enough for that case.

Also would be better do not use raw types as long as possible.

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,279 @@
+					result.set(job);
+				}
+			}
+		}, Duration.ofSeconds(2));
```
I think this should be configurable

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,279 @@
+
+	private boolean sync = false;
+
+	public BigQueryFileMessageHandler(BigQuery bigQuery, TaskScheduler taskScheduler) {
```
There is ` getTaskScheduler()` and `setTaskScheduler()` in the super class.
So, this one must not be required from the ctor.

dzou replied:
Ok,  this is done, but I am afraid users will forget to do this so I added an assert as well:

```
Assert.notNull(getTaskScheduler(),
				"You must set a Task Scheduler for BigQueryFileMessageHandler before using "
						+ "by calling BigQueryFileMessageHandler.setTaskScheduler(taskScheduler).");
```

artembilan replied:
Please, see that getter source code: it falls back into `IntegrationContextUtils.getTaskScheduler(this.beanFactory)` and that one is always there in the application context when we use Spring Integration, e.g. via `@EnableIntegration`.
Or Spring Boot auto-configuration like in most of our tests and samples...

dzou replied:
Ok, I updated the message so it mentions you must use Spring Integration or call setTaskScheduler; this will protect against users instantiating this class directly with `new BigQueryFileMessageHandler(..)`

I also updated the tests to autowire objects instead of use direct instantiation.

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQuerySpringMessageHeaders.java:
```diff
@@ -0,0 +1,50 @@
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.bigquery.integration.outbound;
```
I think this is wrong place for the headers abstraction.
Consider to move it into the previous `integration` package

dzou commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,279 @@
+	protected ListenableFuture handleRequestMessage(Message<?> message) {
+		String datasetName =
+				message.getHeaders().containsKey(BigQuerySpringMessageHeaders.DATASET_NAME)
+						? message.getHeaders().get(BigQuerySpringMessageHeaders.DATASET_NAME, String.class)
```
Sorry, could you clarify what you would like me to do here? Currently this would default to expression evaluation if no header is set right?

artembilan replied:
Well, essentially it is very easy to to provide an expression to evaluate against headers: `headers[gcp_bigquery_dataset_name]`, so, there is even no reason to fallback to one or another.
On the other hand end-user may not provide neither of them. And we should fail that *an expression must evaluate to null*.
We prefer a single place of truth: Therefore I suggest to have only an expression, but with default to be evaluated against headers, even if it returns `null`. That's already not our responsibility to fallback to other places. End-user should just know that expression must return something. Default it or custom - is out of our control.

Let me know if this is still not clear.
(Sorry for confuse if that)

dzou replied:
Done, i set the default expression to read from headers and used expression as source of truth.

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -62,75 +65,97 @@
+
+		ExpressionParser expressionParser = new SpelExpressionParser();
+		this.datasetNameExpression = expressionParser.parseExpression(
+				String.format(HEADER_FORMAT, BigQuerySpringMessageHeaders.DATASET_NAME));
```
As I said: this must be a `FunctionExpression`. No reason to go a parsing and SpEL processing way: we just can call a `Function<Message, String>` instead.

artembilan commented on spring-cloud-gcp-bigquery/src/test/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandlerIntegrationTests.java:
```diff
@@ -73,8 +73,9 @@ public void setup() {
-		this.messageHandler = new BigQueryFileMessageHandler(this.bigquery, this.taskScheduler);
+		this.messageHandler = new BigQueryFileMessageHandler(this.bigquery);
 		this.messageHandler.setWriteDisposition(WriteDisposition.WRITE_TRUNCATE);
+		this.messageHandler.setTaskScheduler(this.taskScheduler);
```
The `taskScheduler` must be destroyed in the end of test. Otherwise some orphan threads are going to stay yet.

dzou replied:
If I rely on the autowired instance through using `@EnableIntegration` then destroying it will not be necessary correct?

artembilan replied:
Correct. That was a comment on outdated commit. Sorry for the noise. Although now you know what to do in test in the future anyway 😉

artembilan commented on spring-cloud-gcp-bigquery/src/test/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandlerIntegrationTests.java:
```diff
@@ -145,10 +152,35 @@ public void testLoadFile_cancel() {
+	static class BigQueryTestConfiguration {
+
+		@Bean
+		ThreadPoolTaskScheduler taskScheduler() {
```
There is a default `taskScheduler` bean created by the Spring Integration infrastructure. You don't need this as long as you use `@EnableIntegration` on the `@Configuration` class

artembilan commented on spring-cloud-gcp-bigquery/pom.xml:
```diff
@@ -0,0 +1,27 @@
+  <dependencies>
+    <dependency>
+      <groupId>org.springframework.integration</groupId>
+      <artifactId>spring-integration-core</artifactId>
```
Why have you resolved this conversation, since I don't see `<optional>true</optional>` in this dependency?
Although it looks like you have placed it accidentally at the  `google-cloud-bigquery`.
For me logically must be opposite...

dzou replied:
oops, thanks for your Vigilance

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,310 @@
+	 * @param tableNameExpression the SpEL expression used to evaluate the table name
+	 */
+	public void setTableNameExpression(Expression tableNameExpression) {
+		this.tableNameExpression = tableNameExpression;
```
Why does this come without `Assert.notNull()` ?

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,310 @@
+	 * @param formatOptions the format of the data file being loaded
+	 */
+	public void setFormatOptions(FormatOptions formatOptions) {
+		this.formatOptionsExpression = new FunctionExpression<Message>(m -> formatOptions);
```
This must be a `ValueExpression` instead. No reason in an intermediate `Function`

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,310 @@
+	 * @param formatOptionsExpression the SpEL expression used to evaluate the {@link FormatOptions}
+	 */
+	public void setFormatOptionsExpression(Expression formatOptionsExpression) {
+		this.formatOptionsExpression = formatOptionsExpression;
```
Is it OK to allow `null` here as well?

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,310 @@
+	 * @param writeDisposition whether data should be appended or truncated to the BigQuery table
+	 */
+	public void setWriteDisposition(WriteDisposition writeDisposition) {
+		this.writeDisposition = writeDisposition;
```
An `Assert.notNull()`, too ?

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,310 @@
+	 * @param timeout the {@link Duration} timeout to wait for a file to load
+	 */
+	public void setTimeout(Duration timeout) {
+		this.timeout = timeout;
```
I guess `null` is going to be bad here as well

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,310 @@
+	 * @param pollInterval the {@link Duration} poll interval for BigQuery job status polling
+	 */
+	public void setJobPollInterval(Duration pollInterval) {
+		this.jobPollInterval = pollInterval;
```
DITTO

artembilan commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,310 @@
+					message, "Failed to initialize the BigQuery write job in message handler: " + this);
+		}
+
+		Assert.notNull(getTaskScheduler(),
```
Please,  move this assert into `doInit()`. This is really message-independent and would be great do not have extra logic on each requested message.

dzou replied:
Hmm, is there a better place to put this? I feel like if `doInit()` is called, they are likely already using Spring Integration. This warning message is more targeted to the folks who directly instantiate the object like `handler = new BigQueryFileMessageHandler(bigquery)`; and not creating it through a Spring bean.

dzou replied:
I moved it to the `doInit` in the latest commit - I suppose since the evaluation context initialization is there the class would be broken if the user didn't call that to begin with....

artembilan replied:
This is not what we need to advertise. It is really wrong to use most of those components outside of Spring Context.
More over such a `MessageHandler` object is a part of Spring Integration flow and it is hard to imaging a proper use-case not wiring it into the flow and use just a standalone instance.
So, please, move it into `doInit()` and remove that `Spring Integration` message.
I would even say let's live without such an assert at all! This component should not be used outside of Spring Integration.
You indeed need to adjust your unit tests, but that's all. Just keep in mind that there is no purpose of this class outside if Spring at all

dzou replied:
Ok sounds good, I have removed the assert message. The tests are now autowiring the instance of the message handler instead of instantiating it with `new ...`

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,312 @@
+		this.bigQuery = bigQuery;
+
+		this.datasetNameExpression =
+				new FunctionExpression<Message>(
```
[optional] Would it be easier to read with a `headerExpression()` helper method?

dzou replied:
Hmm, I think i prefer the explicit style; I considered the helper method approach but preferred this to minimize a level of indirection since it is not too much extra code.

artembilan replied:
I even prefer to have this kind of initialization at the place of property definition instead of ctor.
But that's already IMHO...

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,312 @@
+	}
+
+	/**
+	 * Sets the BigQuery dataset name to use.
```
Mention in javadoc that it's mutually exclusive with `setDatasetNameExpression()`? Or that whichever is last will clobber. If you do it, then same for table name and format options.

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,312 @@
+	}
+
+	/**
+	 * Sets whether BigQuery should attempt to autodetect the schema of the data when loading
```
mention default behavior?

dzou replied:
Added defaults in the docs for the more stable options like `sync`, `autodetect`, etc. but I omitted them for the timeouts because I felt that there is a chance the defaults get changed in the future and the doc will get out of sync.

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,312 @@
+	 * @param sync whether {@link BigQueryFileMessageHandler} should wait synchronously for jobs to
+	 * 		complete
+	 */
+	public void setSync(boolean sync) {
```
Sync is a bit of an overloaded abbreviation -- it could mean "synchronize" as in, "load changes to make the same" or "synchronous". Maybe "setBlocking".

artembilan replied:
Well, this is what we use in other similar places 
So, for consistency I would insist to have it like this 😄

elefeint replied:
Whoops. Withdrawing comment.

dzou commented on spring-cloud-gcp-bigquery/src/test/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandlerIntegrationTests.java:
```diff
@@ -55,6 +57,7 @@
 
 @RunWith(SpringRunner.class)
 @SpringBootTest
+@DirtiesContext(classMode = ClassMode.AFTER_EACH_TEST_METHOD)
```
Right so for this one it's because we Autowire the `BigQueryFileMessageHandler` instance, and in one test I set the dataset/table expressions to something explicit, in another test I just rely on the default.

Because I set it in the first test it modifies the state of the message handler bean so I wanted to reset it after each test.

elefeint commented on spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/outbound/BigQueryFileMessageHandler.java:
```diff
@@ -0,0 +1,312 @@
+					message, "Failed to initialize the BigQuery write job in message handler: " + this);
+		}
+
+		// Prepare the polling task for the ListenableFuture result returned to end-user
```
I'd take lines 255 through 275 into a helper method "watchProgress" or "watchUploadStatus" or something that returns `SettableListenableFuture<Job> result`. `handleListenableMessage()` is getting really long.
